File: libs/language/src/lib/merge/processors/SequentialMergingProcessor.ts
import { BlueNode } from '../../model';
import { NodeProvider } from '../../NodeProvider';
import { MergingProcessor } from '../MergingProcessor';

/**
 * A MergingProcessor that executes multiple processors in sequence
 */
export class SequentialMergingProcessor implements MergingProcessor {
  private readonly mergingProcessors: MergingProcessor[];

  /**
   * Creates a new SequentialMergingProcessor with the given processors
   * @param mergingProcessors - Array of processors to execute in sequence
   */
  constructor(mergingProcessors: MergingProcessor[]) {
    this.mergingProcessors = mergingProcessors;
  }

  /**
   * Processes all contained processors in sequence
   */
  process(
    target: BlueNode,
    source: BlueNode,
    nodeProvider: NodeProvider,
  ): BlueNode {
    return this.mergingProcessors.reduce(
      (currentTarget, processor) =>
        processor.process(currentTarget, source, nodeProvider),
      target,
    );
  }

  preserveSource(
    target: BlueNode,
    source: BlueNode,
    nodeProvider: NodeProvider,
  ): BlueNode | undefined {
    for (const processor of this.mergingProcessors) {
      const preserved = processor.preserveSource?.(
        target,
        source,
        nodeProvider,
      );
      if (preserved !== undefined) {
        return preserved;
      }
    }
    return undefined;
  }

  shouldPreserveSource(source: BlueNode, nodeProvider: NodeProvider): boolean {
    return this.mergingProcessors.some((processor) =>
      processor.shouldPreserveSource?.(source, nodeProvider),
    );
  }

  /**
   * Post-processes all contained processors in sequence
   */
  postProcess(
    target: BlueNode,
    source: BlueNode,
    nodeProvider: NodeProvider,
  ): BlueNode {
    return this.mergingProcessors.reduce((currentPostTarget, processor) => {
      if (processor.postProcess) {
        return processor.postProcess(currentPostTarget, source, nodeProvider);
      }
      return currentPostTarget;
    }, target);
  }
}

File: references/java-sdk/src/main/java/blue/language/merge/processor/SequentialMergingProcessor.java
package blue.language.merge.processor;

import blue.language.NodeProvider;
import blue.language.model.Node;
import blue.language.merge.MergingProcessor;
import blue.language.merge.NodeResolver;

import java.util.List;

public class SequentialMergingProcessor implements MergingProcessor {

    private final List<MergingProcessor> mergingProcessors;

    public SequentialMergingProcessor(List<MergingProcessor> mergingProcessors) {
        this.mergingProcessors = mergingProcessors;
    }

    @Override
    public void process(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        mergingProcessors.forEach(e -> e.process(target, source, nodeProvider, nodeResolver));
    }

    @Override
    public void postProcess(Node target, Node source, NodeProvider nodeProvider, NodeResolver nodeResolver) {
        mergingProcessors.forEach(e -> e.postProcess(target, source, nodeProvider, nodeResolver));
    }
}
